src/Week5/HW5.hs

```diff
@@ -0,0 +1,178 @@
+  return result
+-- * or, more specifically, O(VT) where V = num victims, T = num transactions
+
+-- helper
+isIn :: Eq a => [a] ->  a -> Bool
+isIn [] _ = False
+isIn (x:xs) a
+  |a == x    = True
+  |otherwise = isIn xs a
+  
+-- O(NlogN) version --sort victims and transactions and then filter out transactions in O(N) time :) 
+getBadTs :: FilePath -> FilePath -> IO (Maybe [Transaction])
+getBadTs vi tr = do
+  victims <- parseFile vi
+  trx <- parseFile tr
+  let result = case (trx, victims) of
+                    (Nothing, _)     -> Nothing
+                    (_, Nothing)     -> Nothing
+                    (Just t, Just v) -> Just $ filterTrx (DL.sortBy (compare `on` tid) t) (DL.sort v)
+  return result
+
+-- helper (assumes input is sorted)
+filterTrx :: [Transaction] -> [TId] -> [Transaction]
+filterTrx [] _ = []
+filterTrx _ [] = []
+filterTrx (tr:trs) (i:ids) 
+  |tid tr == i = tr : filterTrx trs ids
+  |tid tr < i  = filterTrx trs (i:ids)
+  |tid tr > i  = filterTrx trs ids
+
+-- Exercise 5 -----------------------------------------
+-- For every name, keep track of how much money that person has gained (or lost) as a result of bad transactions
+
+getFlow :: [Transaction] -> Map String Integer
+getFlow []     = Map.empty
+getFlow (x:xs) = Map.insertWith (+) (to x) (amount x) $ Map.insertWith (+) (from x) (- amount x) $ getFlow xs
+-- the `from` function is (+) (-amount) instead of (-) (amount) to ensure that if the key is not yet present in the map, in which case no function is applied, the value inserted will still be negative
+
+-- Exercise 6 -----------------------------------------
+-- The criminal is the person who got the most money: find him/her
+
+getCriminal :: Map String Integer -> String
+getCriminal mp = helper 0 "no-one" $ Map.toList mp
+  where helper _ name []                 = name
+        helper maxi name ((name',val):xs) = if val > maxi then helper val name' xs
+                                            else helper maxi name xs
+
+-- Exercise 7 -----------------------------------------
+-- Generate a new list of transactions to undo the money transfer
+
+undoTs :: Map String Integer -> [TId] -> [Transaction]
+undoTs mp tids = helper (sortDesc (getPayers mp)) (sortDesc (getPayees mp)) tids 
+  where helper [] _ _ = []
+        helper _ [] _ = []
+        helper _ _ [] = []
+        helper (pr:prs) (pe:pes) (i:ids)
+          |snd pr == 0 = helper prs (pe:pes) (i:ids)
+          |snd pe == 0 = helper (pr:prs) pes (i:ids)
+          |otherwise   = 
+                         let payer = fst pr in
+                         let debit = snd pr in
+                         let payee = fst pe in
+                         let credit = snd pe in
+                         let mini = min credit $ abs debit in
+                         let newTr = Transaction {from = payee, to = payer, amount = mini, tid = i} in
+                         newTr : helper ((payer, debit + mini):prs) ((payee, credit - mini):pes) ids
```

Comment:
Just so you know you can avoid the fst/snd calls by using nested pattern matches.

Also note the @ sign, which is called "as pattern". It you give a name to a variable while also matching its structure and possibly giving name to the components. See [As patterns](https://www.haskell.org/tutorial/patterns.html)

```suggestion
undoTs :: Map String Integer -> [TId] -> [Transaction]
undoTs mp = helper (sortDesc (getPayers mp)) (sortDesc (getPayees mp)) 
  where helper [] _ _ = []
        helper _ [] _ = []
        helper _ _ [] = []
        helper (pr@(payer,debit):prs) (pe@(payee,credit):pes) (i:ids)
          | debit == 0 = helper prs (pe:pes) (i:ids)
          | credit == 0 = helper (pr:prs) pes (i:ids)
          | otherwise = 
              let mini = min credit $ abs debit
                  newTr = Transaction {from = payee, to = payer, amount = mini, tid = i}
              in newTr : helper ((payer, debit + mini):prs) ((payee, credit - mini):pes) ids
```

Reply:
that's indeed a lot better! Thanks